pkg/connection/connection.go

```diff
@@ -38,46 +49,97 @@ func New(opts Options, creds Credentials) *ApiConnection {
 	return &ApiConnection{Options: opts, Credentials: creds}
 }
 
-func (conn ApiConnection) GetRequest(verb string, path string, body any) (*http.Request, error) {
-	b, err := json.Marshal(body)
+func (conn ApiConnection) BuildRequest(verb string, path string, body any) (*http.Request, error) {
+	bodyData, err := json.Marshal(body)
 	if err != nil {
 		return nil, err
 	}
 
-	req, err := http.NewRequest(verb, conn.Options.Url, bytes.NewReader(b))
+	request, err := http.NewRequest(verb, conn.Options.URL, bytes.NewReader(bodyData))
 	if err != nil {
 		return nil, err
 	}
-	req.URL.Path = path
+	request.URL.Path = path
 
-	return req, nil
+	conn.setupGenericHeaders(request)
+
+	return request, nil
 }
 
-func (conn ApiConnection) Do(req *http.Request) ([]byte, error) {
-	tr := http.DefaultTransport.(*http.Transport).Clone()
-	// TODO: note that we need to get the RootCAs thingie from the old connection.
-	tr.TLSClientConfig = &tls.Config{InsecureSkipVerify: conn.Options.Secure}
-	tr.Proxy = conn.Options.Proxy
-	httpclient := &http.Client{Transport: tr, Timeout: 60 * time.Second}
+func (conn ApiConnection) Do(request *http.Request) (int, []byte, error) {
+	token, tokenErr := conn.Credentials.Token()
+	if tokenErr != nil {
+		return 0, []byte{}, tokenErr
+	}
+	request.Header.Add("System-Token", token)
 
-	// TODO: add headers (except auth)
+	response, doErr := conn.setupHTTPClient().Do(request)
+	if doErr != nil {
+		return 0, nil, doErr
+	}
+	defer response.Body.Close()
+	code := response.StatusCode
 
-	resp, err := httpclient.Do(req)
-	if err != nil {
-		return nil, err
+	// Update the credentials from the new system token.
+	token = response.Header.Get("System-Token")
+	if err := conn.Credentials.UpdateToken(token); err != nil {
+		return code, nil, err
 	}
-	defer resp.Body.Close()
 
-	// TODO: handle system token
-	// TODO: handle success/bad code
+	if !successCode(response.StatusCode) {
+		msg := parseError(response.Body)
+		return code, nil, fmt.Errorf("API error: %v (code: %v)", msg, response.StatusCode)
+	}
 
-	resBody, err := io.ReadAll(resp.Body)
-	if err != nil {
-		return nil, err
+	data, readErr := io.ReadAll(response.Body)
+	if readErr != nil {
+		return code, nil, readErr
 	}
-	return resBody, nil
+	return code, data, nil
 }
 
 func (conn ApiConnection) GetCredentials() Credentials {
 	return conn.Credentials
 }
+
+func (conn ApiConnection) setupGenericHeaders(request *http.Request) {
+	version := strings.Split(strings.TrimSpace(rawVersion), "~")[0]
+	userAgent := fmt.Sprintf("%s/%s/%s", version, conn.Options.AppName, conn.Options.Version)
```

Comment:
The library will send the `user-agent` as: `library-version/name of app using the library/version of the app`